### ISSUE TITLE
Class_Terminate does not fire in tB for certain scenario (but does in VB6)

### ISSUE THREAD

**Issue by Kr00l** [WaynePhillipsEA/twinbasic#687]
**Describe the bug**
The below example will not fire Class_Terminate in the TestClass but will in VB6.
That let me concludes that something is wrong in tB.

**To Reproduce**
```
Module MainModule

    Public Sub Main()
        With New TestClass
        ' Class_Terminate will not be called in tB but does in VB6.
        End With
    End Sub
	
End Module

Class TestClass
    Implements IDummyClass
    Private This As CDummyClass
        
    Private Sub Class_Initialize()
    Set This = New CDummyClass
    This.Bind Me
    End Sub
    Private Sub Class_Terminate()
    Debug.Print "This should fire."
    End Sub
    
End Class

Class CDummyClass
    Private PropOwner As IDummyClass

    Private Sub Class_Terminate()
    Set PropOwner = Nothing
    End Sub
    Public Sub Bind(ByVal This As IDummyClass)
    Set PropOwner = This
    End Sub

End Class

Class IDummyClass
End Class
```

**Expected behavior**
That Class_Terminate fires in the TestClass, like in VB6.

**Screenshots**
n/a

**Desktop (please complete the following information):**
 - OS: [e.g. Windows 10]
 - twinBASIC compiler version [e.g. 0.13.91]

**Additional context**
n/a

**Comment by WaynePhillipsEA** [comment]
Nope, this doesn't fire in VB6 either, because of the circular reference (TestClass holds reference to CDummyClass which itself holds a back reference to TestClass).

**Comment by Kr00l** [comment]
> Nope, this doesn't fire in VB6 either, because of the circular reference (TestClass holds reference to CDummyClass which itself holds a back reference to TestClass).

I was sure yesterday. Will check again very soon.

**Comment by Kr00l** [comment]
Ok, I tested again in VB6 and it works. So there is something wrong in tB.

Below a screenshot from VB6 with all the components.
![image](https://user-images.githubusercontent.com/33756666/150735263-9669e4b1-8139-4ecc-8304-5bcb83497b15.png)

**Comment by Kr00l** [comment]
I know now what's going on.

```
Public Sub Main()
    With New TestClass
    ' Class_Terminate will not be called in tB but does in VB6.
    End With
    
    MsgBox "ok"
    
End Sub
```

When showing the MsgBox "ok" in VB6 the debug.print "This should fire." is not yet shown.
So VB6 is resolving this circular reference upon shutdown. tB does not resolve this on shutdown causing missing cleanup code.

**Comment by Kr00l** [comment]
Maybe this is interconnected with #197

**Comment by EduardoVB** [comment]
This has to be with the tear down that VB6 does before ending the program.

VB6 does it in an arbitrary order that have lead to some problems sometime.
For problems I mean: when you rely that the class terminates procedures will fire in certain order. VB6 destroys the classes instances sometimes in another order.

**Comment by EduardoVB** [comment]
It is like it destroys all classes that are still loaded and fires the Class_Terminate events of everyone before doing it. That's good, it is better than nothing.

It must be in some order, but may not be what you expect.
I mean: perhaps it does it in the reverse order that they were created, or by some index, IDK, but not according the parent/child relationship of the logic of the program that you could want or need.

**Comment by EduardoVB** [comment]
It needs to happen because the program code didn't release all these instances explicitly.
Sometimes it occurs unwillingly by error (or error in logic) and sometimes because there is no other way (at least in VB6).

We should have an event at app level or something like that that tells us that the program is about to end. So we could do all the clean-up properly.

**Comment by WaynePhillipsEA** [comment]
I wonder how we might solve that to make it deterministic though... could be difficult

**Comment by EduardoVB** [comment]
> I wonder how we might solve that to make it deterministic though... could be difficult

I have been thinking in the past about this issue, and how it should had done properly in VB6. That was my thinking about at the time, it was in the pre-tB era.

It seems not easy, but I remember at the time I thought some logic that could have been taken to try to guess the best logic in the tear down.
Now I don't remember 🤔 what I thought (like 4 years ago).

**Comment by EduardoVB** [comment]
Anyway having the Class_Terminate events fired in any order like VB6 does is much better than nothing. It can solve most of the cases.

**Comment by EduardoVB** [comment]
In exe programs the developer has more control, and should release the circular references explicitly.
But it is more complex in components, when you don't know when the host program is going to end.

More if it has and End statement, or the Stop button was pressed in the IDE when the host program was running.

**Comment by Kr00l** [comment]
Maybe VB6 resolves this based on it's IConnectionPoint-Containers and it's COM-Eventmechanism for classes ?

**Comment by EduardoVB** [comment]
When you have objects contained in another object, you would expect to terminate first the childs and then the parents and so on.

But that's not what often happens in VB6.

Sometimes the Class_Terminate of a parent object would fire first, then some child...
I don't say I experienced a lot with that, but I saw it.

**Comment by Kr00l** [comment]
How does it behave in VB5 ? (I can't test)
Because VB5 does not have the IConnectionPoint-Containers. The class destruction is very fast in VB5.
So I guess it's not related to that.

**Comment by WaynePhillipsEA** [comment]
BTW, it didn't show for me when I tested it because I ran Main from the immediate window

**Comment by Kr00l** [comment]
When VB shuts down it decrements all refcounts to 0 of all COM objects it "knows" of. The order is not known.
But since the order is unknown I can't imagine somebody is dependending of when what terminates.
I think the main dependency is just it terminates at all and not when, so that cleaning can occur. That a class needs to clean itself prior the other class feels exotic.

So if wayne finds a certain order logic I would recommend to just go with that. At least for the beginning until somebody may report something exotic, if any.

**Comment by bclothier** [comment]
For the circular references, one alternative path to consider is supporting weak references. That way, the cleanup is easy to do and we won't have problems with the shutdown and can take any arbitrary order in disposing of objects. That probably can be decorated as an attribute so that a parent and child object are free to create circular references to each other and indicate to the compiler that one or other can go on without the other. 

It might be of interest to note that in .NET, they use a directed graph for their garbage collection which help them dispose of circular references. I do not know whether that can be extrapolated for tB without making it managed.

**Comment by wqweto** [comment]
Object lifetime tracking in VB6 seems to have considerable performance implications. I was aware of this tear-down slowness but never attributed it to the actual zombie shutdown the runtime does on process shutdown.

In our apps we’ve never depended on runtime taking care of our circular deps and always handle weakrefs at precise points in our code in accordance to instance lifetimes (so that we don’t hydrate a weakref to disposed instance). 

My point is that VB6 tried to solve circular deps and failed. There is no need to emulate something that does not work and pay the performance price on cleanup esp. for the large number of small objects use-case.

**Comment by WaynePhillipsEA** [comment]
I have a plan for fixing this, and the performance impact will be negligible.

**Comment by jpbro** [comment]
> Object lifetime tracking in VB6 seems to have considerable performance implications. I was aware of this tear-down slowness but never attributed it to the actual zombie shutdown the runtime does on process shutdown.

Yeah, VB6 tear-down is abysmal. 

[It reared its ugly head in this recent thread over at vbforums.com](https://www.vbforums.com/showthread.php?894991-RESOLVED-JSON-Decoding-of-RC5-RC6). The long and short of it:

VB6 took ~5 **minutes** to tear-down 300,000 vbRichClient5 cCollection objects. I brought the RC5 code over to tB and tear-down of 300,000 TBRC5.cCollection objects took ~250 **milliseconds**! So huge gains there already. If @WaynePhillipsEA's
plan for dealing with circular references turns out to have a negligible impact, then we're in double-win territory :)

**Comment by EduardoVB** [comment]
> But since the order is unknown I can't imagine somebody is dependending of when what terminates.

Sometimes you need to do it in an ordered manner for freeing GDI or some other type of objects.
And since the order is arbitrary, you need to do some juggleries for not getting a crash and for really freeing objects from memory.

**Comment by bclothier** [comment]
Regarding the weak references, others might be interested in [this code review thread](https://codereview.stackexchange.com/questions/245660/simulated-weakreference-class)

**Comment by fafalone** [comment]
This issue is still present as of Beta 438.
